src/uct/cuda/cuda_copy/cuda_copy_md.c

```diff
@@ -202,14 +207,94 @@ UCS_PROFILE_FUNC(ucs_status_t, uct_cuda_copy_mem_dereg,
     return UCS_OK;
 }
 
-static ucs_status_t uct_cuda_copy_mem_alloc(uct_md_h md, size_t *length_p,
-                                            void **address_p,
-                                            ucs_memory_type_t mem_type,
-                                            unsigned flags,
-                                            const char *alloc_name,
-                                            uct_mem_h *memh_p)
+static ucs_status_t
+uct_cuda_copy_mem_alloc_fabric(uct_cuda_copy_md_t *md,
+                               uct_cuda_copy_alloc_handle_t *alloc_handle)
+{
+#if HAVE_CUDA_FABRIC
+    CUmemAllocationProp prop    = {};
+    CUmemAccessDesc access_desc = {};
+    ucs_status_t status;
+    CUdevice cu_device;
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuCtxGetDevice(&cu_device));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    prop.type                 = CU_MEM_ALLOCATION_TYPE_PINNED;
+    prop.requestedHandleTypes = CU_MEM_HANDLE_TYPE_FABRIC;
+    prop.location.type        = CU_MEM_LOCATION_TYPE_DEVICE;
+    prop.location.id          = cu_device;
+
+    if (md->granularity == SIZE_MAX) {
+        status = UCT_CUDADRV_FUNC_LOG_ERR(cuMemGetAllocationGranularity(
+                &md->granularity, &prop, CU_MEM_ALLOC_GRANULARITY_MINIMUM));
+        if (status != UCS_OK) {
+            return status;
+        }
+    }
+
+    alloc_handle->length = ucs_align_up(alloc_handle->length, md->granularity);
```

Comment:
seems that we would align up to MD granularity also if fabric allocation fails and we fallback to coda pinned allocation
therefore IMO the function uct_cuda_copy_mem_alloc_fabric should not accept alloc_handle structure, but only length, and return the address and generic_handle. or accept the length, and return alloc_handle structure only when successful.